prov/rxm/src/rxm_conn.c

```diff
@@ -1379,6 +1412,92 @@ static void *rxm_conn_progress(void *arg)
 	return NULL;
 }
 
+static void *rxm_conn_atomic_progress(void *arg)
+{
+	struct rxm_ep *rxm_ep = container_of(arg, struct rxm_ep, util_ep);
+	struct rxm_fabric *rxm_fabric;
+	struct fid *fids[2];
+	struct pollfd fds[2];
+	struct rxm_msg_eq_entry *entry;
+	int poll_required;
+	int nfids;
+	int ret;
+
+	assert(rxm_ep->msg_eq);
+	rxm_fabric = container_of(rxm_ep->util_ep.domain->fabric,
+				  struct rxm_fabric, util_fabric);
+	entry = calloc(1, RXM_MSG_EQ_ENTRY_SZ);
+	if (!entry) {
+		FI_WARN(&rxm_prov, FI_LOG_EP_CTRL,
+			"Unable to allocate memory!\n");
+		return NULL;
+	}
+
+	fids[0] = &rxm_ep->msg_eq->fid;
+	ret = fi_control(fids[0], FI_GETWAIT, (void *) &fds[0].fd);
+	if (ret) {
+		FI_WARN(&rxm_prov, FI_LOG_EP_CTRL,
+			"Unable to get MSG_EP EQ WAIT_FD %d\n", ret);
+		goto exit;
+	}
+	fds[0].events = POLLIN;
+	nfids = 1;
+
+	FI_DBG(&rxm_prov, FI_LOG_EP_CTRL,
+	       "Starting CM conn thread with atomic AUTO_PROGRESS\n");
+
+	while (1) {
+		ofi_ep_lock_acquire(&rxm_ep->util_ep);
```

Comment:
I think if we use rxm_ep->cmap->lock here, that can serialize access to rxm_ep->msg_cq even when FI_THREAD_SAFE is not needed.

Reply:
If we take the rxm_ep->cmap->lock in the non thread-safe scenario, can't the application still access the rxm_ep->msg_cq by initiating progress via calls to the data transfer/CQ API while the CM progress thread is performing fi_trywait() (which will access the underlying msg_cq as well)?

Reply:
If the app specifies FI_THREAD_DOMAIN and uses manual progress, cmap lock would be no-op but only one thread should access MSG_CQ at a time as the progress is manual. In case of auto-progress cmap lock would be available and should protect access to MSG CQ. In fact EP lock would a no-op in case of FI_THREAD_DOMAIN and auto-progress.

I think we would want both EP and cmap lock here.

Reply:
Would the behavior be correct having EP lock a no-op with auto-progress, regardless of the threading level?  Since we're adding a progress thread, the application threading level doesn't matter as much.

Reply:
EP lock would still be required when accessing other data structures like recv queue, unexpected queue, etc regardless of the progress.